### PR TITLE
Create benchmarks for Merkle Mountain Range

### PR DESCRIPTION
**Pull request by mikethetike**
## Description

Creates a few benchmarks for the Merkle Mountain Range structure. I'm not sure which methods on the MMR are used in the rest of the project, so I tried to guess the ones that are important. I think we can probably remove the `pub` on most of the methods to make the interface a bit clearer. I added multiple test case sizes to see what kind of order of complexity each method ran in, but most of them ran in _O(n)_ time

I also added a file to document how to build the solution. I put it in the `docs` folder because I've seen it stored there on a few other blockchain projects.

## Motivation and Context

Connected to #166

## How Has This Been Tested?

The benchmarks were run locally. It doesn't look like any of the other benchmarks are currently being run on the CI box.

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [ ] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [x] Documentation

## Checklist:

* [x] I'm merging against the `development` branch
* [x] I ran `cargo-fmt --all` before pushing
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

### CODE REVIEW

**Comment by CjS77** [comment]
You likely have to rebase/merge against `development` to get tests to pass

File: infrastructure/merklemountainrange/benches/bench.rs [hunk 1]
```diff
@@ -0,0 +1,169 @@
+#[macro_use]
+extern crate criterion;
+#[macro_use]
```
Comment (SWvheerden) [inline infrastructure/merklemountainrange/benches/bench.rs]: this not required with 2018 rust

File: infrastructure/merklemountainrange/benches/bench.rs [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+extern crate serde_derive;
```
Comment (SWvheerden) [inline infrastructure/merklemountainrange/benches/bench.rs]: serde derive is now included in the normal serde crate, just add the feature derive

File: infrastructure/merklemountainrange/benches/bench.rs [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+                .add_database(&"mmr_mmr_checkpoints".to_string())
+                .add_database(&"mmr_mmr_objects".to_string())
+                .add_database(&"mmr_init".to_string())
+                .build()
+                .unwrap();
+
+            b.iter(|| {
+                for _ in 0..size {
+                    mmr.append(vec![H(OsRng.next_u64())]).unwrap();
+                    mmr.apply_state(&mut store).unwrap();
```
Comment (SWvheerden) [inline infrastructure/merklemountainrange/benches/bench.rs]: when saving to disc, there are two function: `apply_state` and `checkpoint`.  You have to checkpoint before applying state.  

Checkpoint groups all unsaved changes into change ledge for the lack of a better word (in our context this is like a block). Apply state saves all unsaved checkpoints to disc and makes it kinda permanent. If you call rewind or rest you will go back to the latest applied state

File: infrastructure/merklemountainrange/benches/bench.rs [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+                mmr.push(i).unwrap();
+                items.push(mmr.get_object_hash(x).unwrap());
+            }
+
+            b.iter(|| {
+                let index = OsRng.gen_range(0, size);
+                dbg!(index);
+                let proof = mmr.get_hash_proof(&items[index]);
+                dbg!(proof.len());
+                // dbg!(&proof);
```
Comment (SWvheerden) [inline infrastructure/merklemountainrange/benches/bench.rs]: did you intend to leave these dbg's here?

**Comment by CjS77** [comment]
Technically, the .md files should go in a different PR

**Comment by CjS77** [comment]
Also, I reckon build.md should be in the README, but that's a nit

**Pull request merged**